File: Diagrammatize.Design.Pattern/code/python/Strategy/Sample/WinningStrategy.py
import random

from Strategy import Strategy
from Hand import Hand

class WinningStrategy(Strategy):
    def __init__(self, seed):
        self.random = random.Random(seed)
        self.won = False
        self.prevHand = None

    def nextHand(self):
        if (not self.won):
            self.prevHand = Hand.getHand(self.random.randint(0, 2))
        return self.prevHand

    def study(self, win):
        self.won = win

File: Diagrammatize.Design.Patterns/code/cxx/Strategy/Sample/WinningStrategy.cpp
#include "WinningStrategy.hpp"

WinningStrategy::WinningStrategy(int seed): random(seed) {
}

Hand WinningStrategy::nextHand() {
    if (!won) {
        prevHand = Hand::getHand(random.randomInt(0, 2));
    }
    return prevHand;
}

void WinningStrategy::study(bool win) {
    won = win;
}
